### ISSUE TITLE
Worker 1 stops processing completely after Worker 2 starts stealing leases.

### ISSUE THREAD

**Issue by ewestern** [500px/kinesis-stream#2]
Hi,
Thanks for making this library available.
I've run into an issue when testing this out on a large-ish stream. The issue comes into play when starting a second worker on the application. The new worker (Worker 2) will start to steal leases, as expected. But as soon as the old worker (Worker 1) notices, work on all other shards begins to grind to a halt. Worker 1 will eventually shutdown as soon as CheckpointConfig.timeout has elapsed. Worker 2 will continue on just fine and will eventually end up with all the leases.

To my untrained eye, it looks like there may be some race condition that comes into play when a lease is stolen, and the Dispatcher gets shutdown earlier than expected.

I'm including a slightly curated section of debug output relating to a single shard (the first one the ends up timing out), starting right when Worker 2 steals its first lease. I'd appreciate any insight you have here.

====

```
Feb 06 13:40:25 ip-172-31-21-209 java[17635]: 13:40:25.585 DEBUG [LeaseRenewer-0011] s.a.k.l.d.DynamoDBLeaseRefresher - Renewing lease with key shardId-000000001212
Feb 06 13:40:25 ip-172-31-21-209 java[17635]: 13:40:25.591 DEBUG [LeaseRenewer-0011] s.a.k.l.d.DynamoDBLeaseRenewer - Worker i-09400d30a3c7beef3 successfully renewed lease with key shardId-000000001212
Feb 06 13:40:28 ip-172-31-21-209 java[17635]: 13:40:28.905 DEBUG [LeaseRenewer-0005] s.a.k.l.d.DynamoDBLeaseRefresher - Renewing lease with key shardId-000000001212
Feb 06 13:40:29 ip-172-31-21-209 java[17635]: 13:40:29.015 DEBUG [LeaseRenewer-0005] s.a.k.l.d.DynamoDBLeaseRenewer - Worker i-09400d30a3c7beef3 successfully renewed lease with key shardId-000000001212
Feb 06 13:40:29 ip-172-31-21-209 java[17635]: 13:40:29.506 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT20.852S ago.  Not submitting new task.
Feb 06 13:40:30 ip-172-31-21-209 java[17635]: 13:40:30.507 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT21.853S ago.  Not submitting new task.
Feb 06 13:40:31 ip-172-31-21-209 java[17635]: 13:40:31.508 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT22.854S ago.  Not submitting new task.
Feb 06 13:40:32 ip-172-31-21-209 java[17635]: 13:40:32.206 DEBUG [LeaseRenewer-0003] s.a.k.l.d.DynamoDBLeaseRefresher - Renewing lease with key shardId-000000001212
Feb 06 13:40:32 ip-172-31-21-209 java[17635]: 13:40:32.213 DEBUG [LeaseRenewer-0003] s.a.k.l.d.DynamoDBLeaseRenewer - Worker i-09400d30a3c7beef3 successfully renewed lease with key shardId-000000001212
Feb 06 13:40:32 ip-172-31-21-209 java[17635]: 13:40:32.509 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT23.855S ago.  Not submitting new task.
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: 13:40:33.597 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT0.096S ago.  Not submitting new task.
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: 13:40:33.616 DEBUG [ShardRecordProcessor-0046] s.a.kinesis.lifecycle.ProcessTask - Calling application processRecords() with 5506 records from shardId-000000001212
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: 13:40:33.627 DEBUG [default-akka.actor.default-dispatcher-50] s.a.k.c.ShardRecordProcessorCheckpointer - Checkpointing shardId-000000001212, token d5afb3c2-32c7-44e9-b447-c825597b676a at specific extended sequence number {SequenceNumber: 49590114569102449425719911162763455643313037808593030082,SubsequenceNumber: 626}
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: 13:40:33.627 DEBUG [default-akka.actor.default-dispatcher-50] s.a.k.c.ShardRecordProcessorCheckpointer - Setting shardId-000000001212, token d5afb3c2-32c7-44e9-b447-c825597b676a checkpoint to {SequenceNumber: 49590114569102449425719911162763455643313037808593030082,SubsequenceNumber: 626}
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: 13:40:33.627 DEBUG [default-akka.actor.default-dispatcher-50] s.a.k.l.d.DynamoDBLeaseRefresher - Updating lease Lease(leaseKey=shardId-000000001212, leaseOwner=i-09400d30a3c7beef3, leaseCounter=43808, concurrencyToken=d5afb3c2-32c7-44e9-b447-c825597b676a, lastCounterIncrementNanos=49669894780451, checkpoint={SequenceNumber: 49590114569102449425719911162763455643313037808593030082,SubsequenceNumber: 626}, pendingCheckpoint=null, ownerSwitchesSinceCheckpoint=0, parentShardIds=[shardId-000000001055, shardId-000000001054])
Feb 06 13:40:33 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:33.633] [default-akka.actor.default-dispatcher-50] [akka://default/user/tracker-i-094/shardId-000000001212] Checkpointed Successfully: shardId-000000001212 is at 8593030082
Feb 06 13:40:34 ip-172-31-21-209 java[17635]: 13:40:34.598 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT1.097S ago.  Not submitting new task.
Feb 06 13:40:35 ip-172-31-21-209 java[17635]: 13:40:35.528 DEBUG [LeaseRenewer-0007] s.a.k.l.d.DynamoDBLeaseRefresher - Renewing lease with key shardId-000000001212
Feb 06 13:40:35 ip-172-31-21-209 java[17635]: 13:40:35.535 DEBUG [LeaseRenewer-0007] s.a.k.l.d.DynamoDBLeaseRenewer - Worker i-09400d30a3c7beef3 successfully renewed lease with key shardId-000000001212
Feb 06 13:40:35 ip-172-31-21-209 java[17635]: 13:40:35.599 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT2.098S ago.  Not submitting new task.
Feb 06 13:40:36 ip-172-31-21-209 java[17635]: 13:40:36.601 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT3.1S ago.  Not submitting new task.
Feb 06 13:40:37 ip-172-31-21-209 java[17635]: 13:40:37.602 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT4.101S ago.  Not submitting new task.
Feb 06 13:40:38 ip-172-31-21-209 java[17635]: 13:40:38.603 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Previous PROCESS task still pending for shard shardId-000000001212 since PT5.102S ago.  Not submitting new task.
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: 13:40:56.904 DEBUG [ForkJoinPool.commonPool-worker-9] s.a.kinesis.lifecycle.ShutdownTask - Invoking shutdown() for shard shardId-000000001212, concurrencyToken d5afb3c2-32c7-44e9-b447-c825597b676a. Shutdown reason: LEASE_LOST
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: 13:40:56.904 DEBUG [ForkJoinPool.commonPool-worker-9] s.a.kinesis.lifecycle.ShutdownTask - Record processor completed shutdown() for shard shardId-000000001212
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.907] [default-akka.actor.default-dispatcher-10] [akka://default/user/tracker-i-094/shardId-000000001212] Shutting down tracker shardId-000000001212
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-40] [akka://default/user/tracker-i-094/shardId-000000001212] shardId-000000001212 Tracker started using Dispatcher[akka.actor.default-dispatcher]
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-26] [akka://default/user/tracker-i-094/shardId-000000001212] Message [px.kinesis.stream.consumer.checkpoint.ShardCheckpointTrackerActor$Process] from Actor[akka://default/temp/$a] to Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570] was not delivered. [1] dead letters encountered. If this is not an expected behavior, then [Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-26] [akka://default/user/tracker-i-094/shardId-000000001212] Message [px.kinesis.stream.consumer.checkpoint.ShardCheckpointTrackerActor$Process] from Actor[akka://default/temp/$b] to Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570] was not delivered. [2] dead letters encountered. If this is not an expected behavior, then [Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
Feb 06 13:40:57 ip-172-31-21-209 java[17635]: 13:40:57.905 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Shutdown(shardId-000000001212): Lease lost triggered.
Feb 06 13:40:57 ip-172-31-21-209 java[17635]: 13:40:57.905 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Shutdown(shardId-000000001212): Subscriber cancelled.
Feb 06 13:40:58 ip-172-31-21-209 java[17635]: 13:40:58.906 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Shutdown(shardId-000000001212): Lease lost triggered.
Feb 06 13:40:58 ip-172-31-21-209 java[17635]: 13:40:58.906 DEBUG [KinesisScheduler-0000] s.a.kinesis.lifecycle.ShardConsumer - Shutdown(shardId-000000001212): Subscriber cancelled.
Feb 06 13:40:58 ip-172-31-21-209 java[17635]: 13:40:58.906 DEBUG [KinesisScheduler-0000] s.a.kinesis.coordinator.Scheduler - Removed consumer for shardId-000000001212 as lease has been lost
Feb 06 13:45:56 ip-172-31-21-209 java[17635]: 13:45:56.934 ERROR [ForkJoinPool-1-worker-7] i.n.a.DataLakeLoader$Loader - process took longer than: Timeout(5 minutes) for shardId-000000001212
Feb 06 13:48:31 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:48:31.149] [default-akka.actor.default-dispatcher-7] [akka://default/user/tracker-i-094/shardId-000000001212] Shutting down tracker shardId-000000001212
```

**Comment by ukayani** [comment]
Hi @ewestern, 

Could you provide a few more details?

- How many shards do you have in total?
- Are the 2 workers running on the same machine?
- Are you using the default LeaseManagementConfig?
- What parallelism have you set for the marking of messages as processed?

Just to clarify, the logs above are from Worker 1?

From what i see in the logs (assuming they are from worker 1):
Worker 1 is : `i-09400d30a3c7beef3`
> Worker 1 is processing on shard-1212, taking quite a lot of time processing a single batch of records.

> Worker 1 is periodically renewing its lease (normal) for shard-1212 and eventually the tracker for shard-1212 requests a checkpoint  after having processed some records. It successfully checkpoints.

```
[akka://default/user/tracker-i-094/shardId-000000001212] Checkpointed Successfully: shardId-000000001212 is at 8593030082
```

> Eventually worker 2 steals the lease for shard-1212 at which point the scheduler for Worker 1 notifies the record processor (shard-1212), telling it that the lease is lost and then proceeding to shut it down.

```
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: 13:40:56.904 DEBUG [ForkJoinPool.commonPool-worker-9] s.a.kinesis.lifecycle.ShutdownTask - Invoking shutdown() for shard shardId-000000001212, concurrencyToken d5afb3c2-32c7-44e9-b447-c825597b676a. Shutdown reason: LEASE_LOST
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: 13:40:56.904 DEBUG [ForkJoinPool.commonPool-worker-9] s.a.kinesis.lifecycle.ShutdownTask - Record processor completed shutdown() for shard shardId-000000001212
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.907] [default-akka.actor.default-dispatcher-10] [akka://default/user/tracker-i-094/shardId-000000001212] Shutting down tracker shardId-000000001212
```

The last log above is normal, the tracker associated with shard-1212 is correctly shut down since the corresponding record processor has shut down.

> We then see that a new tracking actor is started for shard-1212

```
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-40] [akka://default/user/tracker-i-094/shardId-000000001212] shardId-000000001212 Tracker started using Dispatcher[akka.actor.default-dispatcher]
```

My guess is that there is an in-flight message to the tracker for shard-1212 that is delivered after it shuts down, this causes a new tracker to be created for that shard.

> Some in flight process messages (these are sent when you mark a record as processed) are delivered to dead letters between the time that the actor shuts down and is recreated (not restarted).

```
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-26] [akka://default/user/tracker-i-094/shardId-000000001212] Message [px.kinesis.stream.consumer.checkpoint.ShardCheckpointTrackerActor$Process] from Actor[akka://default/temp/$a] to Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570] was not delivered. [1] dead letters encountered. If this is not an expected behavior, then [Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
Feb 06 13:40:56 ip-172-31-21-209 java[17635]: [INFO] [02/06/2019 13:40:56.910] [default-akka.actor.default-dispatcher-26] [akka://default/user/tracker-i-094/shardId-000000001212] Message [px.kinesis.stream.consumer.checkpoint.ShardCheckpointTrackerActor$Process] from Actor[akka://default/temp/$b] to Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570] was not delivered. [2] dead letters encountered. If this is not an expected behavior, then [Actor[akka://default/user/tracker-i-094/shardId-000000001212#-554709570]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```
 
> Because these messages were delivered to dead letters instead of the intended actor, the Future that is returned by the `markProcessed` method on the Records will timeout after 5min. If you have a parallelism of 1, you will see that your stream will essentially halt, waiting for this future to complete. This is probably the reason you see the stream for Worker 1 "grind to a halt" when it loses the lease. Once the future fails, your entire stream will fail, resulting in the worker shutting down.

I'm still a little confused as to why the messages get delivered to dead letters instead of to the newly created shard tracker actor. Indeed, there appears to be a race condition here when leases are lost and messages are being marked processed. 

Could you enable debug logs for the `px.kinesis.stream.consumer` package? I'd like to get a little more insight into what exactly is going on.

Also, are you experiencing this issue consistently? In the setup we have, when the worker dies, it gets restarted (container restarts). It should eventually steal leases without issue but would cause a lot of fighting between workers for them to eventually "come to terms".

Thanks for identifying the issue.

In the meantime, when I get a chance, I will look into addressing the race condition with shard tracker creation.

**Comment by ewestern** [comment]
Thanks for looking into this.
To your questions:
* 90 shards in total
* Two workers are on different machines.
* I am using the default lease management config, with the exception that I'm setting maxLeasesForWorker to a default of 100.
* As for parallelism, the pipeline is forked into ~ 16 substreams (using .groupBy). Substreams of course run in parallel, but within each substream, messages are batched for processing. After each batch is processed, the batch is traversed and Future computations are submitted asynchronously.
`Flow[() => Future[Done]].mapAsync(16)(_.toVector.traverse(_.apply()))`
* Correct, the logs are for Worker 1.
* Yes, the issue is consistent. 

Happy to try to get more log output. Will circle back when I get a chance.

**Comment by ewestern** [comment]
> Because these messages were delivered to dead letters instead of the intended actor, the Future that is returned by the `markProcessed` method on the Records will timeout after 5min. If you have a parallelism of 1, you will see that your stream will essentially halt, waiting for this future to complete. This is probably the reason you see the stream for Worker 1 "grind to a halt" when it loses the lease. Once the future fails, your entire stream will fail, resulting in the worker shutting down.

I think I see the problem. Because I am doing a traversal of the Seq[Future], that means that the future of the entire batch won't resolve until the last future of the Seq. Since one of those is the message that went to DLQ, the entire batch gets halted.  Breaking up the batch and checkpointing records individually does seem to remedy the problem. Thanks for this!

**Comment by ukayani** [comment]
@ewestern , Thanks for the info.

Indeed that is a large stream, 90 shards 😄.

>I think I see the problem. Because I am doing a traversal of the Seq[Future], that means that the future of the entire batch won't resolve until the last future of the Seq. Since one of those is the message that went to DLQ, the entire batch gets halted. Breaking up the batch and checkpointing records individually does seem to remedy the problem. Thanks for this!

Yes, doing individual messages should help in your case however this race condition still exists.

I believe I've identified the problem and i'll hopefully release a fix soon.

**Comment by nbenns** [comment]
Openning again until the release is done.

**Comment by ukayani** [comment]
@ewestern I just released `0.1.4`, it should be available on Maven Central in about 20min. Please let me know if it fixes the issue you see.

**Comment by ewestern** [comment]
Great, I'll let you know @ukayani

**Comment by ewestern** [comment]
This has fixed the issue. Thanks @ukayani !

**Comment by ukayani** [comment]
@ewestern , Glad it fixed the issue. I did make a blooper in that release though, left an info log statement that will probably pollute your logs. I've released `0.1.5` which does not have that log message. 

Cheers.

**Comment by ukayani** [comment]
Closing as it has been addressed by #3
